synapse/lib/cmdr.py

```diff
@@ -9,36 +20,595 @@
+        Get the input string to parse.
+
+        Args:
+            prompt (str): Optional string to use as the prompt. Otherwise self.cmdprompt is used.
+
+        Notes:
+            This fires a 'cli:getinput' event prior to getting the
+            input string from the user.
+
+        Returns:
+            str: A string to process.
+        '''
+        return s_cli.Cli.get_input(self, prompt)
+
+    def runCmdLine(self, line):
+        '''
+        Run a single command line on the remote CLI.
+
+        Args:
+            line (str): Line to execute.
+
+        Examples:
+            Execute the 'woot' command with the 'help' switch:
+
+                cli.runCmdLine('woot --help')
+
+        Notes:
+            This generates a random guid and sets the proxy event
+            'syn:cliserver:outp:<guid>' to call self._onPrintf() in order
+            to capture the output sent from the remote server.
+
+        Returns:
+            object: Arbitrary data from the cmd class.
+        '''
+
+        self.fire('cli:cmd:run', line=line)
+
+        iden = s_common.guid()
+        evtname = 'syn:cliserver:outp:%s' % iden
+
+        with self.prox.onWith(evtname, self._onPrintf):
+            ret = self.prox.runCmdLine(iden, line)
+
+        self.fire('cli:cmd:ret', line=line, ret=ret)
+
+        return ret
+
+    def runCmdLoop(self):
+        '''
+        Run commands from a user in an interactive fashion until fini() or EOFError is raised.
+
+        Notes:
+
+            In the event that a disconnect is detected with the Proxy, this
+            will prompt the user that a disconnect has occured and subsequent
+            command execution will attempt a reconnect.
+        '''
+        import readline
+        readline.read_init_file()
+
+        while not self.isfini:
+
+            # FIXME history / completion
+
+            try:
+
+                line = self.get_input()
+                if not line:
+                    continue
+
+                line = line.strip()
+                if not line:
+                    continue
+
+                if self.disconnect_evt.is_set():
+                    if line.lower() == self._quit_cmd:
+                        s_cli.CmdQuit.runCmdOpts(self, None)
+
+                    mesg = 'Disconnected - enter "%s" to exit your session or wait for a reconnect to occur.' % \
+                           self._quit_cmd
+                    self.printf(mesg)
+                    continue
+
+                self.runCmdLine(line)
+
+            except KeyboardInterrupt as e:
+                self.printf('<ctrl-c>')
+
+            except s_common.NoCurrSess as e:
+                self.printf('Current session has expired on the server.')
+                self._reconnect()
+
+            except (s_common.CliFini, EOFError) as e:
+                self.fini()
+
+            except Exception as e:
+                s = traceback.format_exc()
+                self.printf(s)
+
+class CliServer(s_config.Config):
```

Comment:
I think that it is confusing that I can connect to a cortex or the same cortex managed by a CliServer and get different behavior